src/censored.jl

```diff
@@ -0,0 +1,487 @@
+    if !iszero(prob_lower)
+        v += prob_lower * abs2(lower - μ)
+    end
+    if !iszero(prob_upper)
+        v += prob_upper * abs2(upper - μ)
+    end
+    return v
+end
+
+# this expectation also uses the following relation:
+# 𝔼_{x ~ τ}[-log d(x)] = S[τ] - log P_{x ~ d₀}(l ≤ x ≤ u)
+#   + (P_{x ~ d₀}(x = l) (log P_{x ~ d₀}(x = l) - log P_{x ~ d₀}(x ≤ l)) + 
+#      P_{x ~ d₀}(x = u) (log P_{x ~ d₀}(x = u) - log P_{x ~ d₀}(x ≥ u))
+#   ) / P_{x ~ d₀}(l ≤ x ≤ u),
+# where S[τ] is the entropy of τ.
+
+function entropy(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    if lower === missing
+        log_prob_upper = logccdf(d0, upper)
+        if value_support(typeof(d0)) === Discrete
+            logpu = logpdf(d0, upper)
+            log_prob_upper_inc = logaddexp(log_prob_upper, logpu)
+            xlogx_pu = xexpx(logpu)
+        else
+            log_prob_upper_inc = log_prob_upper
+            xlogx_pu = 0
+        end
+        entropy_bound = -xexpx(log_prob_upper_inc)
+        log_prob_trunc = log1mexp(log_prob_upper)    
+        xlogx_pl = 0
+    elseif upper === missing
+        log_prob_lower_inc = logcdf(d0, lower)
+        if value_support(typeof(d0)) === Discrete
+            logpl = logpdf(d0, lower)
+            log_prob_lower = logsubexp(log_prob_lower_inc, logpl)
+            xlogx_pl = xexpx(logpl)
+        else
+            log_prob_lower = log_prob_lower_inc
+            xlogx_pl = 0
+        end
+        entropy_bound = -xexpx(log_prob_lower_inc)
+        log_prob_trunc = log1mexp(log_prob_lower)    
+        xlogx_pu = 0
+    else
+        log_prob_lower_inc = logcdf(d0, lower)
+        log_prob_upper = logccdf(d0, upper)
+        if value_support(typeof(d0)) === Discrete
+            logpl = logpdf(d0, lower)
+            logpu = logpdf(d0, upper)
+            log_prob_lower = logsubexp(log_prob_lower_inc, logpl)
+            log_prob_upper_inc = logaddexp(log_prob_upper, logpu)
+            xlogx_pl = xexpx(logpl)
+            xlogx_pu = xexpx(logpu)
+        else
+            log_prob_lower = log_prob_lower_inc
+            log_prob_upper_inc = log_prob_upper
+            xlogx_pl = xlogx_pu = 0
+        end
+        entropy_bound = -(xexpx(log_prob_lower_inc) + xexpx(log_prob_upper_inc))
+        log_prob_trunc = log1mexp(logaddexp(log_prob_lower, log_prob_upper))
+    end
+    
+    # truncation contains ~ no probability
+    if log_prob_trunc < log(eps(one(log_prob_trunc)))
+        entropy_trunc = zero(entropy_bound)
+    else
+        entropy_trunc = oftype(entropy_bound, entropy(_to_truncated(d)))
+    end
+    entropy_interval = exp(log_prob_trunc) * entropy_trunc - xexpx(log_prob_trunc) + xlogx_pl + xlogx_pu
+    return entropy_bound + entropy_interval
+end
+
+
+#### Evaluation
+
+function pdf(d::Censored{<:Any,<:Any,T}, x0::Real) where {T}
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    S = Base.promote_eltype(T, x0)
+    x = S(x0)
```

Comment:
Mostly so that a `Float64` result is computed actually with `Float64` precision instead of `Float32` precision.

As mentioned somewhere above, since the bound may not be in the support of the uncensored distribution, but it is in the support of the censored distribution, and only the values of the bounds tell us if that is the case or not, then we need to factor the types of the bounds into the computation of the results.

Reply:
I still don't see why the type of the bounds should matter here. In the lines below we just check if e.g. `x0 == lower` or `x0 == upper`, and in this case we can just evaluate `cdf(d0, x0)` and `ccdf(d0, x0)` etc. The computations themselves don't use `lower` or `upper`.

Reply:
Suppose `d0` has all `Int` parameters and `x` has `Int` eltype, but `lower` and `upper` are `Float32`, while `pdf(d0, x)` for `Int` eltype needs to return a float, it promotes to `Float64`, and while `pdf(d0, x)` for `Float32` eltype returns `Float32`. Without factoring in the types of `lower` and `upper`, the user will get `Float64` return values, when it seems like by providing all `Int` and `Float32` parameters, they should only get `Float32` return values.

Reply:
Since the bounds don't enter the computation, to me it seems the expected result for `Int` parameters and `Int` x value is `Float64`. This is what you would obtain for any other distribution as well. If you want an output of type `Float32` you either have to make the parameters of `d0` `Float32` or `x` a `Float32`.

Reply:
I've dropped all promotions based on bounds.